src/java.base/share/classes/jdk/internal/foreign/abi/AbstractLinker.java

```diff
@@ -101,12 +102,19 @@ public SystemLookup defaultLookup() {
         return SystemLookup.getInstance();
     }
 
-    // Current limitation of the implementation:
-    // We don't support packed structs on some platforms,
-    // so reject them here explicitly
-    private static void checkHasNaturalAlignment(FunctionDescriptor descriptor) {
-        descriptor.returnLayout().ifPresent(AbstractLinker::checkHasNaturalAlignmentRecursive);
-        descriptor.argumentLayouts().forEach(AbstractLinker::checkHasNaturalAlignmentRecursive);
+    // Native linkers do not support sequence or padding layouts. They also do not support "packed" struct layouts, or
+    // struct layouts containing additional padding. Such illegal layouts are ruled out here.
+    private static void checkIsSupportedDescriptor(FunctionDescriptor descriptor) {
+        descriptor.returnLayout().ifPresent(AbstractLinker::checkIsSupportedLayout);
+        descriptor.argumentLayouts().forEach(AbstractLinker::checkIsSupportedLayout);
+    }
+
+    private static void checkIsSupportedLayout(MemoryLayout layout) {
+        if (layout instanceof SequenceLayout || layout instanceof PaddingLayout) {
```

Comment:
Maybe we even want to check for padding layouts when creating the function descriptor. Does padding ever make sense as an argument/return layout?

Reply:
I believe FunctionDescriptor::toMethodType is specified to throw if it sees padding?

Reply:
Yes, I mean reject padding upon construction of the FunctionDescriptor.

Reply:
> Yes, I mean reject padding upon construction of the FunctionDescriptor (e.g. using `of` or `ofVoid`).

Ah... good point! I'll think some more.